### PR TITLE
[JCLOUDS-442] reduces the firewalls created by createNodesInGroup

### PR DESCRIPTION
**Pull request by andreaturli**

### CODE REVIEW

**Comment by danbroudy** [comment]
I agree this is still an issue! Im a little foggy on what is a problematic call, how Jclouds handles it before this PR and how it will handle it afterwards. Ill look into the code a little more.

File: google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/strategy/CreateNodesWithGroupEncodedIntoNameThenAddToSet.java [hunk 1]
```diff
@@ -134,52 +128,51 @@ protected CreateNodesWithGroupEncodedIntoNameThenAddToSet(
...
-    * for all sources on both TCP and UDP protocols.
+    * For each group of nodes, there must be a firewall which opens the requested ports for all sources on both TCP and UDP protocols.
     * @see org.jclouds.googlecomputeengine.features.FirewallApi#patch(String, org.jclouds.googlecomputeengine.options.FirewallOptions)
     */
-   private void getOrCreateFirewalls(GoogleComputeEngineTemplateOptions templateOptions, Network network,
-                                     FirewallTagNamingConvention naming) {
-
+   private void getAndUpdateOrCreateFirewalls(GoogleComputeEngineTemplateOptions templateOptions, Network network,
+                                             String sharedResourceName) {
+      String firewallName = templateOptions.getNetworkName().or(sharedResourceName);
```
Comment (danbroudy) [inline google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/strategy/CreateNodesWithGroupEncodedIntoNameThenAddToSet.java]: Ping. Seeing jclouds-labs-google was not promoted to core if we want to make these kinds of changes now is the time to do it.

**Comment by andreaturli** [comment]
@danbroudy have you had the chance to look at it? I know the PR has been closed but I'm seeing issue with the current implementation. http://pastebin.com/3WxHWj2X 
could you advice on that? thanks!

File: google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/strategy/CreateNodesWithGroupEncodedIntoNameThenAddToSet.java [hunk 2]
```diff
@@ -134,52 +128,51 @@ protected CreateNodesWithGroupEncodedIntoNameThenAddToSet(
...
                  MILLISECONDS).apply(operation);
-         checkState(!operation.get().getHttpError().isPresent(),"Could not create firewall, operation failed" + operation);
+         checkState(!operation.get().getHttpError().isPresent(), "Could not patch firewall, operation failed " + operation);
+      }
+   }
+
+   private Set<Rule> createFirewallRulesFromInboundPorts(int[] inboundPorts) {
+      Set<Rule> rules = Sets.newLinkedHashSet();
+      for (int port : inboundPorts) {
+         rules.add(Rule.permitTcpRule(port));
```
Comment (danbroudy) [inline google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/strategy/CreateNodesWithGroupEncodedIntoNameThenAddToSet.java]: This shouldn't add one rule per port. Rules can specify arrays of ranges. We need some logic to find the minimal number of rules for the given inbound ports. We should specify one rule for Tcp and one rule for Udp and each of those rules should enumerate all the ports.

using the ports form this pastebin as an example: http://pastebin.com/3WxHWj2X
[22, 4369, 6000-7999, 8087, 8093, 8098-8099, 8985] covers all the ports
```
"allowed": [
  {
    "IPProtocol": "tcp",
    "ports":["22", "4369", "6000-7999", "8087", "8093", "8098-8099", "8985"],
  },
  {
    "IPProtocol": "udp",
    "ports": ["22", "4369", "6000-7999", "8087", "8093", "8098-8099", "8985"] ,
  }
]
```

https://cloud.google.com/compute/docs/networking#firewalls_1 is a good resource

**Comment by andreaturli** [comment]
Thanks @danbroudy I hope to have sono the chance to fix that and hopefully also add a good logica to delete firewall rules once the VM gets deleted. I'd really like to have your thoughts once the PR is ready

**Comment by danbroudy** [comment]
Happy to help!

**Pull request closed without merging**